File: app/src/main/java/com/clevertap/segmenttest/CleverTapSegmentApplication.java
package com.clevertap.segmenttest;

import android.app.Application;
import android.app.NotificationManager;
import android.os.Build;
import android.os.Build.VERSION_CODES;
import android.os.Handler;
import android.util.Log;

import androidx.annotation.RequiresApi;

import com.clevertap.android.sdk.ActivityLifecycleCallback;
import com.clevertap.android.sdk.CleverTapAPI;
import com.clevertap.android.sdk.SyncListener;
import com.segment.analytics.Analytics;
import com.segment.analytics.android.integrations.clevertap.CleverTapIntegration;

import org.json.JSONObject;

public class CleverTapSegmentApplication extends Application implements SyncListener {

    private static final String TAG = String.format("%s.%s", "CLEVERTAP", CleverTapSegmentApplication.class.getName());
    private static final String WRITE_KEY = "YOUR_KEY";
    private static final String CLEVERTAP_KEY = "CleverTap";
    public static boolean sCleverTapSegmentEnabled = false;

    private CleverTapAPI clevertap;

    private static Handler handler = null;

    @Override public void onCreate() {
        super.onCreate();

        CleverTapAPI.setDebugLevel(CleverTapAPI.LogLevel.VERBOSE);

        Analytics analytics = new Analytics.Builder(getApplicationContext(), WRITE_KEY)
                .logLevel(Analytics.LogLevel.VERBOSE)
                .use(CleverTapIntegration.FACTORY)
                .build();

        analytics.onIntegrationReady(CLEVERTAP_KEY, new Analytics.Callback<CleverTapAPI>() {
            @Override
            public void onReady(CleverTapAPI instance) {
                ActivityLifecycleCallback.register(CleverTapSegmentApplication.this);
                Log.i(TAG, "analytics.onIntegrationReady() called");
                CleverTapIntegrationReady(instance);
            }
        });

        Analytics.setSingletonInstance(analytics);
    }

    private void CleverTapIntegrationReady(CleverTapAPI instance) {
        instance.enablePersonalization();
        sCleverTapSegmentEnabled = true;
        clevertap = instance;
        clevertap.setSyncListener(this);
        getCleverTapAttributionIdentifier();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CleverTapAPI.createNotificationChannel(getApplicationContext(), "BRTesting", "YourChannelName",
                    "YourChannelDescription",
                    NotificationManager.IMPORTANCE_MAX, true);
        }


    }

    private void getCleverTapAttributionIdentifier() {
        //on initial app install, a call to getCleverTapAttributionIdentifier will return NULL until the sdk is fully initialized
        String cleverTapID = clevertap.getCleverTapAttributionIdentifier();
        if (cleverTapID == null) {
            if (handler == null) {
                handler = new Handler();
            }
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    getCleverTapAttributionIdentifier();
                }
            }, 500);
        } else {
            Log.d("CT_ATTRIBUTION_ID", cleverTapID);
        }
    }

    // SyncListener
    public void profileDidInitialize(String CleverTapID){
        Log.d("CT_PROFILE_INITIALIZED", CleverTapID);
    }
    public void profileDataUpdated(JSONObject updates) {
        Log.d("CT_PROFILE_UPDATES", updates.toString());
    }
}

File: app/src/main/java/com/clevertap/segmenttest/CleverTapSegmentApplication.kt
package com.clevertap.segmenttest

import android.app.Application
import android.app.NotificationManager
import android.os.Build
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.clevertap.android.sdk.CTInboxListener
import com.clevertap.android.sdk.CleverTapAPI
import com.clevertap.android.sdk.SyncListener
import com.segment.analytics.Analytics
import com.segment.analytics.android.integrations.clevertap.CleverTapIntegration
import org.json.JSONObject

class CleverTapSegmentApplication : Application() {
    companion object {
        private const val TAG = "CT_SG_SAMPLE_APP>>"
        const val SOURCE_ANDROID_KEY = "VVsCywzFPA8wbGLyNFxwYnixchTUknJy"
        private const val DESTIN_CLEVERTAP_KEY = "CleverTap"
    }
    var ctCoreApi: CleverTapAPI? = null
        set(value) {
            Log.e(TAG, "ctCoreApi called with field = $field | value: $value")
            field = value
        }

    var segmentAnalytics : Analytics? = null

    override fun onCreate() {
        CleverTapAPI.setDebugLevel(CleverTapAPI.LogLevel.VERBOSE)
        super.onCreate()
        segmentCTIntegration()
    }

    private fun segmentCTIntegration() {
        val analytics: Analytics = Analytics.Builder(applicationContext, SOURCE_ANDROID_KEY)
            .logLevel(Analytics.LogLevel.VERBOSE)
            .use(CleverTapIntegration.FACTORY)
            .build()
        analytics.onIntegrationReady(DESTIN_CLEVERTAP_KEY, ::initCleverTap)  ///<----
        Analytics.setSingletonInstance(analytics)
    }



    private fun initCleverTap(ctInstance: CleverTapAPI?) {
        Log.e(TAG, "initCleverTap() called with: instance = $ctInstance. call onresume to update button")
        ctInstance ?: return
        ctInstance.enablePersonalization()
        ctInstance.syncListener = object : SyncListener {
            override fun profileDidInitialize(CleverTapID: String) { Log.e(TAG, "profileDidInitialize() called with: CleverTapID = $CleverTapID") }
            override fun profileDataUpdated(updates: JSONObject) { Log.e(TAG, "profileDataUpdated() called with: updates = \n ${updates.toString(4)}") }
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CleverTapAPI.createNotificationChannel(applicationContext, "BRTesting", "YourChannelName", "YourChannelDescription", NotificationManager.IMPORTANCE_MAX, true)
        }

        //on initial app install, a call to getCleverTapAttributionIdentifier will return NULL until the sdk is fully initialized
        Log.e(TAG, "CT_ATTRIBUTION_ID: ${ctInstance.cleverTapAttributionIdentifier}")

        ctInstance.initializeInbox()
        ctInstance.ctNotificationInboxListener = object : CTInboxListener {
            override fun inboxDidInitialize() { Log.e(TAG, "inboxDidInitialize() called. call onresume to update button") }
            override fun inboxMessagesDidUpdate() { Log.e(TAG, "inboxMessagesDidUpdate() called. call onresume to update button") }
        }

        ctCoreApi = ctInstance
        segmentAnalytics = Analytics.with(applicationContext)

    }
}
